lightningd/lightningd.c

```diff
@@ -254,46 +409,74 @@ static void init_txfilter(struct wallet *w, struct txfilter *filter)
 	}
 }
 
+/*~ The normal advice for daemons is to move into the root directory, so you
+ * don't prevent unmounting whatever filesystem you happen to start in.
+ *
+ * But we define every path relative to our (~/.lightning) data dir, so we
+ * make sure we stay there.
+ */
 static void daemonize_but_keep_dir(struct lightningd *ld)
 {
 	/* daemonize moves us into /, but we want to be here */
 	const char *cwd = path_cwd(NULL);
 
+	/*~ SQLite3 does NOT like being open across fork(), a.k.a. daemonize() */
 	db_close_for_fork(ld->wallet->db);
 	if (!cwd)
 		fatal("Could not get current directory: %s", strerror(errno));
 	if (!daemonize())
 		fatal("Could not become a daemon: %s", strerror(errno));
 
-	/* Move back: important, since lightning dir may be relative! */
+	/*~ Move back: important, since lightning dir may be relative! */
 	if (chdir(cwd) != 0)
 		fatal("Could not return to directory %s: %s",
 		      cwd, strerror(errno));
 
 	db_reopen_after_fork(ld->wallet->db);
+
+	/*~ Why not allocate cwd off tmpctx?  Probably because this code predates
+	 * tmpctx.  So we free manually here. */
 	tal_free(cwd);
 }
 
+/*~ It's pretty standard behaviour (especially for daemons) to create and
+ * file-lock a pidfile.  This not only prevents accidentally running multiple
+ * daemons on the same database at once, but lets nosy sysadmins see what pid
+ * the currently-running daemon is supposed to be. */
 static void pidfile_create(const struct lightningd *ld)
 {
 	char *pid;
+	int pid_fd;
 
 	/* Create PID file */
 	pid_fd = open(ld->pidfile, O_WRONLY|O_CREAT, 0640);
 	if (pid_fd < 0)
 		err(1, "Failed to open PID file");
 
-	/* Lock PID file */
+	/* Lock PID file: this will stay locked until we exit. */
 	if (lockf(pid_fd, F_TLOCK, 0) < 0)
 		/* Problem locking file */
 		err(1, "lightningd already running? Error locking PID file");
 
-	/* Get current PID and write to PID fie */
+	/*~ Note that tal_fmt() is what asprintf() dreams of being. */
 	pid = tal_fmt(tmpctx, "%d\n", getpid());
+	/*~ CCAN's write_all writes to a file descriptor, looping if necessary
+	 * (which, on a file unlike a socket, is never, for historical UNIX
+	 * reasons).  It also isn't declared with GCC's warn_unused_result
+	 * which write() is when FORTIFY_SOURCE is defined, so we're allowed
+	 * to ignore the result without jumping through hoops. */
 	write_all(pid_fd, pid, strlen(pid));
+
+	/* Leave file open: we close it implicitly when we exit */
```

Comment:
because the os does or parent is closed or how do I have to understand this?

Reply:
OS.  fixed.